<a href="https://colab.research.google.com/github/iampramodyadav/ML/blob/main/Amazon_Fine_Food_CurlWget.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>



```
`# This is formatted as code`
```

# CSV File (Loading data from kaggle via Curlwget)
https://www.kaggle.com/snap/amazon-fine-food-reviews?select=Reviews.csv

https://www.kaggle.com/snap/amazon-fine-food-reviews?select=database.sqlite

In [ ]:
!wget --header="Host: storage.googleapis.com" --header="User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.77 Safari/537.36" --header="Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9" --header="Accept-Language: en-US,en;q=0.9,hi;q=0.8" --header="Referer: https://www.kaggle.com/" "https://storage.googleapis.com/kaggle-data-sets/18/2157/compressed/Reviews.csv.zip?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=gcp-kaggle-com%40kaggle-161607.iam.gserviceaccount.com%2F20210604%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20210604T075708Z&X-Goog-Expires=259199&X-Goog-SignedHeaders=host&X-Goog-Signature=3d01b1beb1c1090a4d933f58b6414e637fa59926f38f8924ba0e40508576af6a7ccb3e1b4dbd8cbdd8565f28f3a6ab79b50077f0afa1983c70b8f06c8925aac664c028e7fd4ea7454d0628693d722259edd0e36076840bebc76653c7b6edd56b81203da8f6d4f46ab1ab56a229d0c3528bdd84811b51bcba4388bef1c00b5b15e3bf2c7ec9d5d92ef64104998e8a4b9627b88949f0b6b627419c51c1f8f1151b09f3e00d15fcb2f850656f70937e6c759c0cd64c47361b851bc2a6ad7cd5a27738fde63f15c7a9000df5cc288f000fe702dae7f3697d1dccb059f83a6d6f02b75542366b08e7e27d419a806a49d1a2ee1200c39b69b90ba7898879c3e3f4f15b" -c -O 'Reviews.csv.zip'

--2021-06-04 07:57:47--  https://storage.googleapis.com/kaggle-data-sets/18/2157/compressed/Reviews.csv.zip?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=gcp-kaggle-com%40kaggle-161607.iam.gserviceaccount.com%2F20210604%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20210604T075708Z&X-Goog-Expires=259199&X-Goog-SignedHeaders=host&X-Goog-Signature=3d01b1beb1c1090a4d933f58b6414e637fa59926f38f8924ba0e40508576af6a7ccb3e1b4dbd8cbdd8565f28f3a6ab79b50077f0afa1983c70b8f06c8925aac664c028e7fd4ea7454d0628693d722259edd0e36076840bebc76653c7b6edd56b81203da8f6d4f46ab1ab56a229d0c3528bdd84811b51bcba4388bef1c00b5b15e3bf2c7ec9d5d92ef64104998e8a4b9627b88949f0b6b627419c51c1f8f1151b09f3e00d15fcb2f850656f70937e6c759c0cd64c47361b851bc2a6ad7cd5a27738fde63f15c7a9000df5cc288f000fe702dae7f3697d1dccb059f83a6d6f02b75542366b08e7e27d419a806a49d1a2ee1200c39b69b90ba7898879c3e3f4f15b
Resolving storage.googleapis.com (storage.googleapis.com)... 64.233.170.128, 108.177.13.128, 172.217.193.128, ...
Connecting to storag

In [ ]:
import pandas as pd
data=pd.read_csv('Reviews.csv.zip')
data.head(5)

Id  ...                                               Text
0   1  ...  I have bought several of the Vitality canned d...
1   2  ...  Product arrived labeled as Jumbo Salted Peanut...
2   3  ...  This is a confection that has been around a fe...
3   4  ...  If you are looking for the secret ingredient i...
4   5  ...  Great taffy at a great price.  There was a wid...

[5 rows x 10 columns]

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 568454 entries, 0 to 568453
Data columns (total 10 columns):
 #   Column                  Non-Null Count   Dtype 
---  ------                  --------------   ----- 
 0   Id                      568454 non-null  int64 
 1   ProductId               568454 non-null  object
 2   UserId                  568454 non-null  object
 3   ProfileName             568438 non-null  object
 4   HelpfulnessNumerator    568454 non-null  int64 
 5   HelpfulnessDenominator  568454 non-null  int64 
 6   Score                   568454 non-null  int64 
 7   Time                    568454 non-null  int64 
 8   Summary                 568427 non-null  object
 9   Text                    568454 non-null  object
dtypes: int64(5), object(5)
memory usage: 43.4+ MB


# Amazon Fine Food Reviews Analysis


Data Source: https://www.kaggle.com/snap/amazon-fine-food-reviews <br>

EDA: https://nycdatascience.com/blog/student-works/amazon-fine-foods-visualization/


The Amazon Fine Food Reviews dataset consists of reviews of fine foods from Amazon.<br>

Number of reviews: 568,454<br>
Number of users: 256,059<br>
Number of products: 74,258<br>
Timespan: Oct 1999 - Oct 2012<br>
Number of Attributes/Columns in data: 10 

Attribute Information:

1. Id
2. ProductId - unique identifier for the product
3. UserId - unqiue identifier for the user
4. ProfileName
5. HelpfulnessNumerator - number of users who found the review helpful
6. HelpfulnessDenominator - number of users who indicated whether they found the review helpful or not
7. Score - rating between 1 and 5
8. Time - timestamp for the review
9. Summary - brief summary of the review
10. Text - text of the review


#### Objective:
Given a review, determine whether the review is positive (Rating of 4 or 5) or negative (rating of 1 or 2).

<br>
[Q] How to determine if a review is positive or negative?<br>
<br> 
[Ans] We could use the Score/Rating. A rating of 4 or 5 could be cosnidered a positive review. A review of 1 or 2 could be considered negative. A review of 3 is nuetral and ignored. This is an approximate and proxy way of determining the polarity (positivity/negativity) of a review.




## Loading the data

The dataset is available in two forms
1. .csv file
2. SQLite Database

In order to load the data, We have used the SQLITE dataset as it easier to query the data and visualise the data efficiently.
<br> 

Here as we only want to get the global sentiment of the recommendations (positive or negative), we will purposefully ignore all Scores equal to 3. If the score id above 3, then the recommendation wil be set to "positive". Otherwise, it will be set to "negative".

# SQL DATABASE (Loading data from kaggle via curlwget and unzip)

https://www.kaggle.com/snap/amazon-fine-food-reviews?select=database.sqlite

In [1]:
!wget --header="Host: storage.googleapis.com" --header="User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.77 Safari/537.36" --header="Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9" --header="Accept-Language: en-US,en;q=0.9,hi;q=0.8" --header="Referer: https://www.kaggle.com/" "https://storage.googleapis.com/kaggle-data-sets/18/2157/compressed/database.sqlite.zip?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=gcp-kaggle-com%40kaggle-161607.iam.gserviceaccount.com%2F20210604%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20210604T080213Z&X-Goog-Expires=259199&X-Goog-SignedHeaders=host&X-Goog-Signature=6b217033240b2f91419db3a91f919dee0d81cab4b7ea7922971d1778f7570bf3d81471998128c31e6f02aad6a7e5f1a2819006babadaf1e2631c8bfeec36175590f85a0aa142d486db7f8bd2f04496cddff2feb0fe815a68ca5cee2ecca8a347718927c16578440eb5ba71e7b65a921ba613a205adc2e6af867b75e67c9e0199e7f8e82764d66f14adc6945e2ee896c305225329f8de8a1d81fec4ccdd6a18d72822204b87780a3beeb4ad52a101727a5dde00cffb61a3564b59659a0198c65d1498af610977b8ba6cc69e3d0fa03d9d051243d1e83b19353d333353e8f7cd61ae109d091ded5435746a4de9dd4b37e982167a23a6ea2bc60b57ad722edb8caf" -c -O 'database.sqlite.zip'

--2021-06-04 12:57:25--  https://storage.googleapis.com/kaggle-data-sets/18/2157/compressed/database.sqlite.zip?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=gcp-kaggle-com%40kaggle-161607.iam.gserviceaccount.com%2F20210604%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20210604T080213Z&X-Goog-Expires=259199&X-Goog-SignedHeaders=host&X-Goog-Signature=6b217033240b2f91419db3a91f919dee0d81cab4b7ea7922971d1778f7570bf3d81471998128c31e6f02aad6a7e5f1a2819006babadaf1e2631c8bfeec36175590f85a0aa142d486db7f8bd2f04496cddff2feb0fe815a68ca5cee2ecca8a347718927c16578440eb5ba71e7b65a921ba613a205adc2e6af867b75e67c9e0199e7f8e82764d66f14adc6945e2ee896c305225329f8de8a1d81fec4ccdd6a18d72822204b87780a3beeb4ad52a101727a5dde00cffb61a3564b59659a0198c65d1498af610977b8ba6cc69e3d0fa03d9d051243d1e83b19353d333353e8f7cd61ae109d091ded5435746a4de9dd4b37e982167a23a6ea2bc60b57ad722edb8caf
Resolving storage.googleapis.com (storage.googleapis.com)... 172.217.203.128, 172.253.123.128, 108.177.11.128, ...
Connecting to s

In [84]:
import numpy as np
import pandas as pd

import sqlite3

In [3]:
!unzip "database.sqlite.zip"

Archive:  database.sqlite.zip
  inflating: database.sqlite         


In [6]:
con = sqlite3.connect('database.sqlite') 
filteredData=pd.read_sql_query("""SELECT * FROM Reviews WHERE Score!=3 LIMIT 5000""",con)

In [7]:
filteredData

Id  ...                                               Text
0        1  ...  I have bought several of the Vitality canned d...
1        2  ...  Product arrived labeled as Jumbo Salted Peanut...
2        3  ...  This is a confection that has been around a fe...
3        4  ...  If you are looking for the secret ingredient i...
4        5  ...  Great taffy at a great price.  There was a wid...
...    ...  ...                                                ...
4995  5423  ...  My baby didn't seem into these dinners, so I t...
4996  5424  ...  This is great!  Organic baby food options - de...
4997  5425  ...  My little guy loves to try new foods..so this ...
4998  5426  ...  We ordered the Earth's best 2nd dinner variety...
4999  5427  ...  My baby loves this food.  At whole foods they ...

[5000 rows x 10 columns]

In [ ]:
%matplotlib inline
import warnings
warnings.filterwarnings("ignore")



import sqlite3
import pandas as pd
import numpy as np
import nltk
import string
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.feature_extraction.text import TfidfVectorizer

from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics import confusion_matrix
from sklearn import metrics
from sklearn.metrics import roc_curve, auc
from nltk.stem.porter import PorterStemmer

import re
# Tutorial about Python regular expressions: https://pymotw.com/2/re/
import string
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer
from nltk.stem.wordnet import WordNetLemmatizer

from gensim.models import Word2Vec
from gensim.models import KeyedVectors
import pickle

from tqdm import tqdm
import os

D:\installed\Anaconda3\lib\site-packages\gensim\utils.py:1197: UserWarning: detected Windows; aliasing chunkize to chunkize_serial
  warnings.warn("detected Windows; aliasing chunkize to chunkize_serial")


In [11]:
def partition(score):
  if score<3:
    return 0
  return 1

In [9]:
ActuScore=filteredData['Score']


In [14]:
positiveNeg=ActuScore.apply(partition)

In [15]:
filteredData['Score']=positiveNeg

In [17]:
filteredData.head(3)

Id  ...                                               Text
0   1  ...  I have bought several of the Vitality canned d...
1   2  ...  Product arrived labeled as Jumbo Salted Peanut...
2   3  ...  This is a confection that has been around a fe...

[3 rows x 10 columns]



---



In [24]:
display=pd.read_sql_query("""
SELECT UserId, ProductId,ProfileName, Time, Score, Text, COUNT(*)
FROM Reviews
GROUP BY UserId
HAVING COUNT(*)>1
""",con)

In [25]:
display

UserId  ... COUNT(*)
0      #oc-R115TNMSPFT9I7  ...        2
1      #oc-R11D9D7SHXIJB9  ...        3
2      #oc-R11DNU2NBKQ23Z  ...        2
3      #oc-R11O5J5ZVQE25C  ...        3
4      #oc-R12KPBODL2B5ZD  ...        2
...                   ...  ...      ...
80663       AZZNK89PXD006  ...        6
80664       AZZTH6DJ0KSIP  ...        3
80665       AZZU1VEO8KUXH  ...        3
80666       AZZU4D6TZ2L6J  ...        2
80667       AZZV9PDNMCOZW  ...        3

[80668 rows x 6 columns]

In [82]:
display['COUNT(*)'].sum()

568454



---
# Rough Work


In [79]:
pd.read_sql_query("""
SELECT UserId,ID, ProductId,ProfileName, Time, Score, Text, COUNT(*)
FROM Reviews
GROUP BY Id

""",con)

UserId  ...  COUNT(*)
0       A3SGXH7AUHU8GW  ...         1
1       A1D87F6ZCVE5NK  ...         1
2        ABXLMWJIXXAIN  ...         1
3       A395BORC6FGVXV  ...         1
4       A1UQRSCLF8GW1T  ...         1
...                ...  ...       ...
568449  A28KG5XORO54AY  ...         1
568450  A3I8AFVPEE8KI5  ...         1
568451  A121AA1GQV751Z  ...         1
568452   A3IBEVCTXKNOH  ...         1
568453  A3LGQPJCZVL9UC  ...         1

[568454 rows x 8 columns]

In [39]:
pd.read_sql_query("""
SELECT *
FROM Reviews
WHERE UserId=='AZZNK89PXD006'
""",con)

Id  ...                                               Text
0   84898  ...  In this package you get three varieties of pop...
1  152692  ...  My vet suggested that a grain-free diet would ...
2  185765  ...  I love Japanese candy, so this package was eag...
3  232226  ...  I've been trying to spice up some of my usual ...
4  314009  ...  For around thirty bucks I was greeted at the d...
5  427048  ...  Regular Chocolate Pocky is good on its own but...

[6 rows x 10 columns]

In [40]:
pd.read_sql_query("""
SELECT *, COUNT(*)
FROM Reviews
WHERE UserId=='AZZNK89PXD006'
""",con)

Id  ... COUNT(*)
0  427048  ...        6

[1 rows x 11 columns]

In [42]:
pd.read_sql_query("""
SELECT *, COUNT(*)
FROM Reviews
""",con)

Id  ... COUNT(*)
0  568454  ...   568454

[1 rows x 11 columns]

In [68]:
test=pd.read_sql_query("""
SELECT UserId,ID, ProductId,ProfileName, Time, Score, Text
FROM Reviews
GROUP BY UserId

""",con)

In [69]:
test

UserId  ...                                               Text
0       #oc-R103C0QSV1DF5E  ...  I have to say I was a little apprehensive to b...
1       #oc-R109MU5OBBZ59U  ...  Received my free K cups as a sample promotion ...
2       #oc-R10LFEMQEW6QGZ  ...  Brooklyn Bean Roastery Blend K-Cups are great ...
3       #oc-R10LT57ZGIB140  ...  The shipment of the dog food was quick.  Howev...
4       #oc-R10UA029WVWIUI  ...  Brooklyn "French Roast" K-Cup Coffee is not on...
...                    ...  ...                                                ...
256054       AZZV9PDNMCOZW  ...  Even though this mix is not as spicy as I anti...
256055       AZZVNIMTTMJH6  ...  i've wanted one of these since i saw it in an ...
256056       AZZY649VYAHQS  ...  All of the chocolate arrived with out melting ...
256057       AZZYCJOJLUDYR  ...  I use this all the time! Good for BBQ ing or p...
256058       AZZZOVIBXHGDR  ...  I ordered this cereal based on the good review...

[256059 rows x 7 columns]

In [81]:
test[test['UserId']=='AZZNK89PXD006']

UserId  ...                                               Text
256039  AZZNK89PXD006  ...  Regular Chocolate Pocky is good on its own but...

[1 rows x 7 columns]

In [75]:
pd.read_sql_query("""
SELECT UserId,ID, ProductId,ProfileName, Time, Score, Text
FROM Reviews

WHERE UserId=='AZZV9PDNMCOZW'
""",con)

UserId  ...                                               Text
0  AZZV9PDNMCOZW  ...  I love this syrup. It is my personal substitut...
1  AZZV9PDNMCOZW  ...  This is a great addition to my Torani syrup. I...
2  AZZV9PDNMCOZW  ...  Even though this mix is not as spicy as I anti...

[3 rows x 7 columns]

In [73]:
pd.read_sql_query("""
SELECT UserId,ID, ProductId,ProfileName, Time, Score, Text
FROM Reviews
GROUP BY UserId
HAVING UserId=='AZZV9PDNMCOZW'
""",con)

UserId  ...                                               Text
0  AZZV9PDNMCOZW  ...  Even though this mix is not as spicy as I anti...

[1 rows x 7 columns]

In [52]:
pd.read_sql_query("""
SELECT UserId, ProductId,ProfileName, Time, Score, Text, COUNT(*)
FROM Reviews

WHERE UserId=='AZZV9PDNMCOZW'
""",con)

UserId  ... COUNT(*)
0  AZZV9PDNMCOZW  ...        3

[1 rows x 7 columns]

# Data Cleaning

In [86]:
#Sorting data according to ProductId in ascending order
sorted_data=filteredData.sort_values('ProductId', axis=0, ascending=True, inplace=False, kind='quicksort', na_position='last')

In [88]:
sorted_data.head(5)

Id  ...                                               Text
2546  2774  ...  Why is this $[...] when the same product is av...
2547  2775  ...  We have used the Victor fly bait for 3 seasons...
1145  1244  ...  I just received my shipment and could hardly w...
1146  1245  ...  This was a really good idea and the final prod...
2942  3204  ...  I'm glad my 45lb cocker/standard poodle puppy ...

[5 rows x 10 columns]

In [91]:
sorted_data.shape

(5000, 10)

In [89]:
#Deduplication of entries
final=sorted_data.drop_duplicates(subset={"UserId","ProfileName","Time","Text"}, keep='first', inplace=False)
final.shape

(4986, 10)

In [93]:
final.head(3)

Id  ...                                               Text
2546  2774  ...  Why is this $[...] when the same product is av...
2547  2775  ...  We have used the Victor fly bait for 3 seasons...
1145  1244  ...  I just received my shipment and could hardly w...

[3 rows x 10 columns]

In [96]:
display= pd.read_sql_query("""
SELECT *
FROM Reviews
WHERE Score != 3 AND Id=44737 OR Id=64422
ORDER BY ProductID
""", con)

display.head()

Id  ...                                               Text
0  64422  ...  My son loves spaghetti so I didn't hesitate or...
1  44737  ...  It was almost a 'love at first bite' - the per...

[2 rows x 10 columns]

In [94]:
final=final[final.HelpfulnessNumerator<=final.HelpfulnessDenominator]

In [95]:
#Before starting the next phase of preprocessing lets see the number of entries left
print(final.shape)

#How many positive and negative reviews are present in our dataset?
final['Score'].value_counts()

(4986, 10)


1    4178
0     808
Name: Score, dtype: int64

#   Text Preprocessing.

Now that we have finished deduplication our data requires some preprocessing before we go on further with analysis and making the prediction model.

Hence in the Preprocessing phase we do the following in the order below:-

1. Begin by removing the html tags
2. Remove any punctuations or limited set of special characters like , or . or # etc.
3. Check if the word is made up of english letters and is not alpha-numeric
4. Check to see if the length of the word is greater than 2 (as it was researched that there is no adjective in 2-letters)
5. Convert the word to lowercase
6. Remove Stopwords
7. Finally Snowball Stemming the word (it was obsereved to be better than Porter Stemming)<br>

After which we collect the words used to describe positive and negative reviews

In [97]:
# printing some random reviews
sent_0 = final['Text'].values[0]
print(sent_0)
print("="*50)

sent_1000 = final['Text'].values[1000]
print(sent_1000)
print("="*50)

sent_1500 = final['Text'].values[1500]
print(sent_1500)
print("="*50)

sent_4900 = final['Text'].values[4900]
print(sent_4900)
print("="*50)

Why is this $[...] when the same product is available for $[...] here?<br />http://www.amazon.com/VICTOR-FLY-MAGNET-BAIT-REFILL/dp/B00004RBDY<br /><br />The Victor M380 and M502 traps are unreal, of course -- total fly genocide. Pretty stinky, but only right nearby.
I recently tried this flavor/brand and was surprised at how delicious these chips are.  The best thing was that there were a lot of "brown" chips in the bsg (my favorite), so I bought some more through amazon and shared with family and friends.  I am a little disappointed that there are not, so far, very many brown chips in these bags, but the flavor is still very good.  I like them better than the yogurt and green onion flavor because they do not seem to be as salty, and the onion flavor is better.  If you haven't eaten Kettle chips before, I recommend that you try a bag before buying bulk.  They are thicker and crunchier than Lays but just as fresh out of the bag.
Wow.  So far, two two-star reviews.  One obviously had no 

## re 
[re use](https://pymotw.com/2/re/)

## from bs4 import BeautifulSoup

In [99]:
import re

In [100]:
# remove urls from text python: https://stackoverflow.com/a/40823105/4084039
sent_0 = re.sub(r"http\S+", "", sent_0)
sent_1000 = re.sub(r"http\S+", "", sent_1000)
sent_150 = re.sub(r"http\S+", "", sent_1500)
sent_4900 = re.sub(r"http\S+", "", sent_4900)

print(sent_0)

Why is this $[...] when the same product is available for $[...] here?<br /> /><br />The Victor M380 and M502 traps are unreal, of course -- total fly genocide. Pretty stinky, but only right nearby.


In [101]:
# https://stackoverflow.com/questions/16206380/python-beautifulsoup-how-to-remove-all-tags-from-an-element
from bs4 import BeautifulSoup

soup = BeautifulSoup(sent_0, 'lxml')
text = soup.get_text()
print(text)
print("="*50)

soup = BeautifulSoup(sent_1000, 'lxml')
text = soup.get_text()
print(text)
print("="*50)

soup = BeautifulSoup(sent_1500, 'lxml')
text = soup.get_text()
print(text)
print("="*50)

soup = BeautifulSoup(sent_4900, 'lxml')
text = soup.get_text()
print(text)

Why is this $[...] when the same product is available for $[...] here? />The Victor M380 and M502 traps are unreal, of course -- total fly genocide. Pretty stinky, but only right nearby.
I recently tried this flavor/brand and was surprised at how delicious these chips are.  The best thing was that there were a lot of "brown" chips in the bsg (my favorite), so I bought some more through amazon and shared with family and friends.  I am a little disappointed that there are not, so far, very many brown chips in these bags, but the flavor is still very good.  I like them better than the yogurt and green onion flavor because they do not seem to be as salty, and the onion flavor is better.  If you haven't eaten Kettle chips before, I recommend that you try a bag before buying bulk.  They are thicker and crunchier than Lays but just as fresh out of the bag.
Wow.  So far, two two-star reviews.  One obviously had no idea what they were ordering; the other wants crispy cookies.  Hey, I'm sorry; b

In [103]:
# https://stackoverflow.com/a/47091490/4084039
import re

def decontracted(phrase):
    # specific
    phrase = re.sub(r"won't", "will not", phrase)
    phrase = re.sub(r"can\'t", "can not", phrase)

    # general
    phrase = re.sub(r"n\'t", " not", phrase)
    phrase = re.sub(r"\'re", " are", phrase)
    phrase = re.sub(r"\'s", " is", phrase)
    phrase = re.sub(r"\'d", " would", phrase)
    phrase = re.sub(r"\'ll", " will", phrase)
    phrase = re.sub(r"\'t", " not", phrase)
    phrase = re.sub(r"\'ve", " have", phrase)
    phrase = re.sub(r"\'m", " am", phrase)
    return phrase

In [104]:
sent_1500 = decontracted(sent_1500)
print(sent_1500)
print("="*50)

Wow.  So far, two two-star reviews.  One obviously had no idea what they were ordering; the other wants crispy cookies.  Hey, I am sorry; but these reviews do nobody any good beyond reminding us to look  before ordering.<br /><br />These are chocolate-oatmeal cookies.  If you do not like that combination, do not order this type of cookie.  I find the combo quite nice, really.  The oatmeal sort of "calms" the rich chocolate flavor and gives the cookie sort of a coconut-type consistency.  Now let is also remember that tastes differ; so, I have given my opinion.<br /><br />Then, these are soft, chewy cookies -- as advertised.  They are not "crispy" cookies, or the blurb would say "crispy," rather than "chewy."  I happen to like raw cookie dough; however, I do not see where these taste like raw cookie dough.  Both are soft, however, so is this the confusion?  And, yes, they stick together.  Soft cookies tend to do that.  They are not individually wrapped, which would add to the cost.  Oh y

In [105]:
#remove words with numbers python: https://stackoverflow.com/a/18082370/4084039
sent_0 = re.sub("\S*\d\S*", "", sent_0).strip()
print(sent_0)

Why is this $[...] when the same product is available for $[...] here?<br /> /><br />The Victor  and  traps are unreal, of course -- total fly genocide. Pretty stinky, but only right nearby.


In [106]:
#remove spacial character: https://stackoverflow.com/a/5843547/4084039
sent_1500 = re.sub('[^A-Za-z0-9]+', ' ', sent_1500)
print(sent_1500)

Wow So far two two star reviews One obviously had no idea what they were ordering the other wants crispy cookies Hey I am sorry but these reviews do nobody any good beyond reminding us to look before ordering br br These are chocolate oatmeal cookies If you do not like that combination do not order this type of cookie I find the combo quite nice really The oatmeal sort of calms the rich chocolate flavor and gives the cookie sort of a coconut type consistency Now let is also remember that tastes differ so I have given my opinion br br Then these are soft chewy cookies as advertised They are not crispy cookies or the blurb would say crispy rather than chewy I happen to like raw cookie dough however I do not see where these taste like raw cookie dough Both are soft however so is this the confusion And yes they stick together Soft cookies tend to do that They are not individually wrapped which would add to the cost Oh yeah chocolate chip cookies tend to be somewhat sweet br br So if you wa

In [107]:
# https://gist.github.com/sebleier/554280
# we are removing the words from the stop words list: 'no', 'nor', 'not'
# <br /><br /> ==> after the above steps, we are getting "br br"
# we are including them into stop words list
# instead of <br /> if we have <br/> these tags would have revmoved in the 1st step

stopwords= set(['br', 'the', 'i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've",\
            "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', \
            'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their',\
            'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', "that'll", 'these', 'those', \
            'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', \
            'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', \
            'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after',\
            'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further',\
            'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more',\
            'most', 'other', 'some', 'such', 'only', 'own', 'same', 'so', 'than', 'too', 'very', \
            's', 't', 'can', 'will', 'just', 'don', "don't", 'should', "should've", 'now', 'd', 'll', 'm', 'o', 're', \
            've', 'y', 'ain', 'aren', "aren't", 'couldn', "couldn't", 'didn', "didn't", 'doesn', "doesn't", 'hadn',\
            "hadn't", 'hasn', "hasn't", 'haven', "haven't", 'isn', "isn't", 'ma', 'mightn', "mightn't", 'mustn',\
            "mustn't", 'needn', "needn't", 'shan', "shan't", 'shouldn', "shouldn't", 'wasn', "wasn't", 'weren', "weren't", \
            'won', "won't", 'wouldn', "wouldn't"])

In [108]:
# Combining all the above stundents 
from tqdm import tqdm

preprocessed_reviews = []
# tqdm is for printing the status bar
for sentance in tqdm(final['Text'].values):
    sentance = re.sub(r"http\S+", "", sentance)
    sentance = BeautifulSoup(sentance, 'lxml').get_text()
    sentance = decontracted(sentance)
    sentance = re.sub("\S*\d\S*", "", sentance).strip()
    sentance = re.sub('[^A-Za-z]+', ' ', sentance)
    # https://gist.github.com/sebleier/554280
    sentance = ' '.join(e.lower() for e in sentance.split() if e.lower() not in stopwords)
    preprocessed_reviews.append(sentance.strip())

100%|██████████| 4986/4986 [00:01<00:00, 2656.09it/s]


In [109]:
preprocessed_reviews[1500]

'wow far two two star reviews one obviously no idea ordering wants crispy cookies hey sorry reviews nobody good beyond reminding us look ordering chocolate oatmeal cookies not like combination not order type cookie find combo quite nice really oatmeal sort calms rich chocolate flavor gives cookie sort coconut type consistency let also remember tastes differ given opinion soft chewy cookies advertised not crispy cookies blurb would say crispy rather chewy happen like raw cookie dough however not see taste like raw cookie dough soft however confusion yes stick together soft cookies tend not individually wrapped would add cost oh yeah chocolate chip cookies tend somewhat sweet want something hard crisp suggest nabiso ginger snaps want cookie soft chewy tastes like combination chocolate oatmeal give try place second order'

<h2><font color='red'>[3.2] Preprocess Summary</font></h2>

In [ ]:
## Similartly you can do preprocessing for review summary also.

# [4] Featurization

## [4.1] BAG OF WORDS

In [110]:
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.feature_extraction.text import TfidfVectorizer

from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics import confusion_matrix
from sklearn import metrics
from sklearn.metrics import roc_curve, auc

In [111]:
#BoW
count_vect = CountVectorizer() #in scikit-learn
count_vect.fit(preprocessed_reviews)
print("some feature names ", count_vect.get_feature_names()[:10])
print('='*50)

final_counts = count_vect.transform(preprocessed_reviews)
print("the type of count vectorizer ",type(final_counts))
print("the shape of out text BOW vectorizer ",final_counts.get_shape())
print("the number of unique words ", final_counts.get_shape()[1])

some feature names  ['aa', 'aahhhs', 'aback', 'abandon', 'abates', 'abbott', 'abby', 'abdominal', 'abiding', 'ability']
the type of count vectorizer  <class 'scipy.sparse.csr.csr_matrix'>
the shape of out text BOW vectorizer  (4986, 12997)
the number of unique words  12997


## [4.2] Bi-Grams and n-Grams.

In [112]:
#bi-gram, tri-gram and n-gram

#removing stop words like "not" should be avoided before building n-grams
# count_vect = CountVectorizer(ngram_range=(1,2))
# please do read the CountVectorizer documentation http://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html
# you can choose these numebrs min_df=10, max_features=5000, of your choice
count_vect = CountVectorizer(ngram_range=(1,2), min_df=10, max_features=5000)
final_bigram_counts = count_vect.fit_transform(preprocessed_reviews)
print("the type of count vectorizer ",type(final_bigram_counts))
print("the shape of out text BOW vectorizer ",final_bigram_counts.get_shape())
print("the number of unique words including both unigrams and bigrams ", final_bigram_counts.get_shape()[1])

the type of count vectorizer  <class 'scipy.sparse.csr.csr_matrix'>
the shape of out text BOW vectorizer  (4986, 3144)
the number of unique words including both unigrams and bigrams  3144


## [4.3] TF-IDF

In [113]:
tf_idf_vect = TfidfVectorizer(ngram_range=(1,2), min_df=10)
tf_idf_vect.fit(preprocessed_reviews)
print("some sample features(unique words in the corpus)",tf_idf_vect.get_feature_names()[0:10])
print('='*50)

final_tf_idf = tf_idf_vect.transform(preprocessed_reviews)
print("the type of count vectorizer ",type(final_tf_idf))
print("the shape of out text TFIDF vectorizer ",final_tf_idf.get_shape())
print("the number of unique words including both unigrams and bigrams ", final_tf_idf.get_shape()[1])

some sample features(unique words in the corpus) ['ability', 'able', 'able find', 'able get', 'absolute', 'absolutely', 'absolutely delicious', 'absolutely love', 'absolutely no', 'according']
the type of count vectorizer  <class 'scipy.sparse.csr.csr_matrix'>
the shape of out text TFIDF vectorizer  (4986, 3144)
the number of unique words including both unigrams and bigrams  3144


## [4.4] Word2Vec

In [120]:
import nltk
from nltk.stem.porter import PorterStemmer

import re
# Tutorial about Python regular expressions: https://pymotw.com/2/re/
import string
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer
from nltk.stem.wordnet import WordNetLemmatizer
from tqdm import tqdm
import os

In [127]:
# using google news word2vectors
from gensim.models import Word2Vec
from gensim.models import KeyedVectors
import pickle

In [121]:
# Train your own Word2Vec model using your own text corpus
i=0
list_of_sentance=[]
for sentance in preprocessed_reviews:
    list_of_sentance.append(sentance.split())

In [140]:
# Using Google News Word2Vectors

# in this project we are using a pretrained model by google
# its 3.3G file, once you load this into your memory 
# it occupies ~9Gb, so please do this step only if you have >12G of ram
# we will provide a pickle file wich contains a dict , 
# and it contains all our courpus words as keys and  model[word] as values
# To use this code-snippet, download "GoogleNews-vectors-negative300.bin" 
# from https://drive.google.com/file/d/0B7XkCwpI5KDYNlNUTTlSS21pQmM/edit
# it's 1.9GB in size.


# http://kavita-ganesan.com/gensim-word2vec-tutorial-starter-code/#.W17SRFAzZPY
# you can comment this whole cell
# or change these varible according to your need

is_your_ram_gt_16g=False
want_to_use_google_w2v = False
want_to_train_w2v = True

if want_to_train_w2v:
    # min_count = 5 considers only words that occured atleast 5 times
    w2v_model=Word2Vec(list_of_sentance,min_count=5,size=50, workers=4)
    print(w2v_model.wv.most_similar('great'))
    print('='*50)
    print(w2v_model.wv.most_similar('worst'))
    
elif want_to_use_google_w2v and is_your_ram_gt_16g:
    if os.path.isfile('GoogleNews-vectors-negative300.bin'):
        w2v_model=KeyedVectors.load_word2vec_format('GoogleNews-vectors-negative300.bin', binary=True)
        print(w2v_model.wv.most_similar('great'))
        print(w2v_model.wv.most_similar('worst'))
    else:
        print("you don't have gogole's word2vec file, keep want_to_train_w2v = True, to train your own w2v ")

[('think', 0.9948400259017944), ('bad', 0.9942811131477356), ('anything', 0.9938587546348572), ('want', 0.9938547611236572), ('excellent', 0.9934669137001038), ('care', 0.9932903051376343), ('feel', 0.9931259751319885), ('pretty', 0.992738664150238), ('quality', 0.9927330613136292), ('overall', 0.9927022457122803)]
[('normal', 0.9994822144508362), ('american', 0.9994388818740845), ('including', 0.9994204044342041), ('beef', 0.9994080662727356), ('turned', 0.9993816614151001), ('kid', 0.9993793368339539), ('unfortunately', 0.999373197555542), ('part', 0.9993698596954346), ('nuts', 0.9993577003479004), ('world', 0.9993560314178467)]


In [134]:
w2v_model.wv.similarity('man','woman')

0.9902055

In [141]:
w2v_model.wv.most_similar('like')

[('taste', 0.9950764179229736),
 ('flavor', 0.9930325746536255),
 ('strong', 0.9884700179100037),
 ('sweet', 0.9882386326789856),
 ('tastes', 0.9862771034240723),
 ('really', 0.9829283952713013),
 ('coffee', 0.9823290109634399),
 ('bitter', 0.9774443507194519),
 ('chocolate', 0.9770838618278503),
 ('hot', 0.9768550992012024)]

In [142]:
w2v_words = list(w2v_model.wv.vocab)
print("number of words that occured minimum 5 times ",len(w2v_words))
print("sample words ", w2v_words[0:50])

number of words that occured minimum 5 times  3817
sample words  ['product', 'available', 'course', 'total', 'pretty', 'stinky', 'right', 'nearby', 'used', 'ca', 'not', 'beat', 'great', 'received', 'shipment', 'could', 'hardly', 'wait', 'try', 'love', 'call', 'instead', 'removed', 'easily', 'daughter', 'designed', 'printed', 'use', 'car', 'windows', 'beautifully', 'shop', 'program', 'going', 'lot', 'fun', 'everywhere', 'like', 'tv', 'computer', 'really', 'good', 'idea', 'final', 'outstanding', 'window', 'everybody', 'asks', 'bought', 'made']


## [4.4.1] Converting text into vectors using wAvg W2V, TFIDF-W2V

#### [4.4.1.1] Avg W2v

In [124]:
# average Word2Vec
# compute average word2vec for each review.
sent_vectors = []; # the avg-w2v for each sentence/review is stored in this list
for sent in tqdm(list_of_sentance): # for each review/sentence
    sent_vec = np.zeros(50) # as word vectors are of zero length 50, you might need to change this to 300 if you use google's w2v
    cnt_words =0; # num of words with a valid vector in the sentence/review
    for word in sent: # for each word in a review/sentence
        if word in w2v_words:
            vec = w2v_model.wv[word]
            sent_vec += vec
            cnt_words += 1
    if cnt_words != 0:
        sent_vec /= cnt_words
    sent_vectors.append(sent_vec)
print(len(sent_vectors))
print(len(sent_vectors[0]))

100%|██████████| 4986/4986 [00:05<00:00, 969.72it/s]

4986
50


#### [4.4.1.2] TFIDF weighted W2v

In [125]:
# S = ["abc def pqr", "def def def abc", "pqr pqr def"]
model = TfidfVectorizer()
model.fit(preprocessed_reviews)
# we are converting a dictionary with word as a key, and the idf as a value
dictionary = dict(zip(model.get_feature_names(), list(model.idf_)))

In [126]:
# TF-IDF weighted Word2Vec
tfidf_feat = model.get_feature_names() # tfidf words/col-names
# final_tf_idf is the sparse matrix with row= sentence, col=word and cell_val = tfidf

tfidf_sent_vectors = []; # the tfidf-w2v for each sentence/review is stored in this list
row=0;
for sent in tqdm(list_of_sentance): # for each review/sentence 
    sent_vec = np.zeros(50) # as word vectors are of zero length
    weight_sum =0; # num of words with a valid vector in the sentence/review
    for word in sent: # for each word in a review/sentence
        if word in w2v_words and word in tfidf_feat:
            vec = w2v_model.wv[word]
#             tf_idf = tf_idf_matrix[row, tfidf_feat.index(word)]
            # to reduce the computation we are 
            # dictionary[word] = idf value of word in whole courpus
            # sent.count(word) = tf valeus of word in this review
            tf_idf = dictionary[word]*(sent.count(word)/len(sent))
            sent_vec += (vec * tf_idf)
            weight_sum += tf_idf
    if weight_sum != 0:
        sent_vec /= weight_sum
    tfidf_sent_vectors.append(sent_vec)
    row += 1

100%|██████████| 4986/4986 [00:36<00:00, 136.37it/s]
